File: problems/15/solution.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        if (size(nums) < 3) {
            return {};
        }
        
        int n = size(nums);
        vector<vector<int>> ans;
        
        sort(begin(nums), end(nums));
        
        for (int i = 0; i < n - 2; i++) {
            if (i and nums[i] == nums[i - 1]) {
                continue;
            }
            
            int j = i + 1, k = n - 1;
            
            while (j < k) {
                int sum = nums[i] + nums[j] + nums[k];
                
                if (sum == 0) {
                    ans.push_back({nums[i], nums[j], nums[k]});
                    
                    while (j < k and nums[j] == nums[j + 1]) {
                        j++;
                    }
                    
                    while (j < k and nums[k] == nums[k - 1]) {
                        k--;
                    }
                    
                    j++;
                    k--;
                } else if (sum > 0) {
                    k--;
                } else {
                    j++;
                }
            }
        }
        
        return ans;
    }
};

File: problems/15/solution.cs
public class Solution {
    public IList<IList<int>> ThreeSum(int[] nums) {
        var ans = new List<int[]>();
        
        Array.Sort(nums);
        
        for (var i = 0; i < nums.Length - 2; i++) {
            if (i > 0 && nums[i] == nums[i - 1]) {
                continue;
            }
            
            var a = nums[i];
            var lo = i + 1;
            var hi = nums.Length - 1;
            
            while (lo < hi) {
                var b = nums[lo];
                var c = nums[hi];
                var curSum = a + b + c;
                
                if (curSum == 0) {
                    ans.Add(new int[]{a, b, c});
                    
                    while (lo < hi && nums[lo] == nums[lo + 1]) {
                        lo++;
                    }
                    
                    while (lo < hi && nums[hi] == nums[hi - 1]) {
                        hi--;
                    }
                    
                    lo++;
                    hi--;
                } else if (curSum > 0) {
                    hi--;
                } else {
                    lo++;
                }
            }
        }
        
        return ans.ToArray();
    }
}